File: LC152MaximumProductSubarray.java
package leetcode;

/**
 * https://leetcode.com/problems/maximum-product-subarray/
 */
public class LC152MaximumProductSubarray {
    public int maxProduct(int[] nums) {
        int max = Integer.MIN_VALUE;
        int cmin = 0, cmax = 0;
        for (int i = 0; i < nums.length; i++) {
            int tmin = nums[i], tmax = nums[i];
            if (cmin != 0) {
                tmin *= cmin;
            }
            if (cmax != 0) {
                tmax *= cmax;
            }
            cmin = tmin < tmax ? tmin : tmax;
            cmin = nums[i] < cmin ? nums[i] : cmin;
            cmax = tmin < tmax ? tmax : tmin;
            cmax = nums[i] < cmax ? cmax : nums[i];
            if (cmax > max) {
                max = cmax;
            }
        }
        return max;
    }

    public static void main(String[] argumets) {
        int[] nums = {2, 3, -2, 4};
        LC152MaximumProductSubarray solution = new LC152MaximumProductSubarray();
        System.out.println(solution.maxProduct(nums));
    }
}

File: algorithms/code/leetcode/lc152_maximum_product_subarray/lc152_maximum_product_subarray.py
#!/usr/bin/env python3
# https://leetcode.com/problems/maximum-product-subarray/

import unittest
from typing import List


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        maximum = float('-inf')
        cmin = 0
        cmax = 0
        for num in nums:
            tmin = num
            tmax = num
            if cmin != 0:
                tmin *= cmin
            if cmax != 0:
                tmax *= cmax
            cmin = tmin if tmin < tmax else tmax
            cmin = num if num < cmin else cmin
            cmax = tmax if tmin < tmax else tmin
            cmax = cmax if num < cmax else num
            if cmax > maximum:
                maximum = cmax
        return maximum


class TestCode(unittest.TestCase):
    def test_2_3_Minus2_4(self):
        nums = [2, 3, -2, 4]
        self.assertEqual(6, Solution().maxProduct(nums))
